src/modules/roc_core/target_gcc/roc_core/attributes.h

```diff
@@ -28,4 +28,18 @@
 #define ROC_ATTR_PRINTF(n_fmt_arg, n_var_arg)                                            \
     __attribute__((format(printf, n_fmt_arg, n_var_arg)))
 
+//! Used to suppress undefined behavior sanitizer for a particular function
+#if defined(__clang__)                                                                   \
+    && (__clang_major__ > 3 || __clang_major__ == 3 && __clang_minor__ >= 8)
+#define ROC_ATTR_NO_SANITIZE_UB __attribute__((no_sanitize("undefined")))
+
+#elif !defined(__clang__) && defined(__GNUC__)                                           \
+    && (__GNUC__ > 4 || __GNUC__ == 4 && __GNUC_MINOR >= 9)
+#define ROC_ATTR_NO_SANITIZE_UB __attribute__((no_sanitize_undefined))
```

Comment:
Can we replace these version checks with something like this? (I didn't test it).

```c
#ifdef __has_feature
 #if __has_feature(memory_sanitizer)
  #if defined(__clang__)
   #define ROC_ATTR_NO_SANITIZE_UB ...
  #elif defined(__GNUC__)
   #define ROC_ATTR_NO_SANITIZE_UB ...
  #endif // defined(__GNUC__)
 #endif // __has_feature(memory_sanitizer)
#endif // __has_feature

#ifndef ROC_ATTR_NO_SANITIZE_UB
 //! Doxygen comment here.
 #define ROC_ATTR_NO_SANITIZE_UB
#endif
```

Reply:
I have tested the suggested approach. But GCC does not support `__has_feature` macro.

Reply:
Oh, really. Can we use __has_attribute(no_sanitize_undefined) for GCC?

Reply:
> Oh, really.

Am I wrong? Does GCC indeed support `__has_feature` macro?



> Can we use __has_attribute(no_sanitize_undefined) for GCC?

Yes.

Reply:
It seems that [the oldest](https://roc-streaming.org/toolkit/docs/portability/supported_platforms.html) supported clang and gcc already has __has_feature and __has_attribute, so we can assume they're always present.

Reply:
A typo: `__has_feature(memory_sanitizer)` -> `__has_feature(undefined_behavior_sanitizer)`

Reply:
> Am I wrong? Does GCC indeed support __has_feature macro?

No, you're right, __has_feature is clang-specific.

Reply:
> It seems that the oldest supported clang and gcc already has __has_feature and __has_attribute, so we can assume they're always present.

Hm, maybe I'm wrong and we need an ifdef: https://gcc.gnu.org/gcc-5/changes.html

Reply:
[This](https://gcc.gnu.org/gcc-5/changes.html) page mentions that `__has_attribute` macro is supported from `gcc-5` onwards.